sdk/src/client/api/block_builder/input_selection/requirement/allotment.rs

```diff
@@ -0,0 +1,157 @@
+        input::{Input, UtxoInput},
+        mana::ManaAllotment,
+        output::Output,
+        payload::{signed_transaction::Transaction, SignedTransactionPayload},
+        signature::Ed25519Signature,
+        unlock::{AccountUnlock, NftUnlock, ReferenceUnlock, SignatureUnlock, Unlock, Unlocks},
+        Error as BlockError,
+    },
+};
+
+impl InputSelection {
+    pub(crate) fn fulfill_allotment_requirement(&mut self) -> Result<Vec<InputSigningData>, Error> {
+        let AutoManaAllotment {
+            issuer_id,
+            reference_mana_cost,
+        } = self
+            .auto_mana_allotment
+            .ok_or(Error::UnfulfillableRequirement(Requirement::Allotment))?;
+
+        let mut inputs = Vec::new();
+        for input in &self.selected_inputs {
+            inputs.push(Input::Utxo(UtxoInput::from(*input.output_id())));
+        }
+
+        let mut builder = Transaction::builder(self.protocol_parameters.network_id())
+            .with_inputs(inputs)
+            .with_outputs(self.outputs.clone());
+
+        if let Some(payload) = &self.payload {
+            builder = builder.with_payload(payload.clone());
+        }
+
+        self.mana_allotments.entry(issuer_id).or_default();
+
+        let transaction = builder
+            .with_context_inputs(self.context_inputs.clone())
+            .with_mana_allotments(
+                self.mana_allotments
+                    .iter()
+                    .map(|(&account_id, &mana)| ManaAllotment { account_id, mana }),
+            )
+            .finish_with_params(&self.protocol_parameters)?;
+
+        let signed_transaction = SignedTransactionPayload::new(transaction, self.transaction_unlocks()?)?;
+
+        let block_work_score = self.protocol_parameters.work_score(&signed_transaction)
+            + self.protocol_parameters.work_score_parameters().block();
+        let required_allotment_mana = block_work_score as u64 * reference_mana_cost;
+
+        // Add the required allotment to the issuing allotment
+        let allotment = self.mana_allotments[&issuer_id];
+        if allotment < required_allotment_mana {
+            self.mana_allotments.insert(issuer_id, required_allotment_mana);
+            let additional_inputs = self.fulfill_mana_requirement()?;
+            // If we needed more inputs to cover the additional allotment mana
+            // then re-add this requirement so we try again
+            if !additional_inputs.is_empty() {
+                self.requirements.push(Requirement::Allotment);
+                return Ok(additional_inputs);
+            }
+        }
+
+        Ok(Vec::new())
+    }
+
+    pub(crate) fn transaction_unlocks(&self) -> Result<Unlocks, Error> {
+        // let transaction_signing_hash = prepared_transaction_data.transaction.signing_hash();
+        let mut blocks = Vec::new();
+        let mut block_indexes = HashMap::<Address, usize>::new();
+
+        // Assuming inputs_data is ordered by address type
```

Comment:
This ordering still needs to happen before L39 where the inputs are set for this tx?